File: plugin/src/main/java/de/s42/tradecycle/TradeCyclePlugin.java
package de.s42.tradecycle;

import de.s42.core.Platform;
import de.s42.paper.PaperPlatform;
import de.s42.spigot.SpigotPlatform;
import de.s42.tradecycle.common.ComparableVersion;
import de.s42.tradecycle.common.VersionCheckTask;
import de.s42.tradecycle.listener.VillagerCycleListener;
import de.s42.tradecycle.service.TradeCycleService;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.event.Listener;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.List;

public class TradeCyclePlugin extends JavaPlugin {

    @Override
    public void onEnable() {
        checkForUpdates();
        saveDefaultConfig();

        Platform platform = getPlatform();
        TradeCycleService service = new TradeCycleService(platform.getMessageService(), getConfig());
        registerStrategies(platform, service, getConfig());
    }

    private void checkForUpdates() {
        new VersionCheckTask(latestVersion -> {
            ComparableVersion currentVersion = new ComparableVersion(getDescription().getVersion());
            if (currentVersion.compareTo(latestVersion) < 0) {
                getLogger().warning("A new version of TradeCycle is available: " + latestVersion.getRawVersion() + ". You are using version: " + currentVersion.getRawVersion());
                getLogger().warning("Download the latest version at: " + VersionCheckTask.DIRECT_DOWNLOAD);
            }
        }).runTaskAsynchronously(this);
    }

    private Platform getPlatform() {
        try {
            Class.forName("com.destroystokyo.paper.PaperConfig");
            return new PaperPlatform();
        } catch (ClassNotFoundException e) {
            return new SpigotPlatform();
        }
    }

    private void registerStrategies(Platform platform, TradeCycleService service, FileConfiguration config) {
        List<Listener> strategies = config.getStringList("strategy").stream()
                .map(s -> {
                    Listener l = TradeCycleService.strategies.get(s);
                    if (l == null) throw new IllegalArgumentException("Invalid strategy: " + s);
                    return l;
                })
                .toList();

        PluginManager pluginManager = getServer().getPluginManager();
        pluginManager.registerEvents(new VillagerCycleListener(service, platform.getOpenInventoryViewFunction(), this), this);
        strategies.forEach(s -> pluginManager.registerEvents(s, this));
    }
}

File: plugin/src/main/kotlin/tr/s42/tradecycle/TradeCyclePlugin.kt
package tr.s42.tradecycle

import tr.s42.core.Platform
import tr.s42.tradecycle.common.ComparableVersion
import tr.s42.tradecycle.common.VersionCheckTask
import tr.s42.tradecycle.listener.VillagerCycleListener
import tr.s42.tradecycle.service.TradeCycleService
import tr.s42.tradecycle.service.TradeCycleService.Companion.strategies
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.event.Listener
import org.bukkit.plugin.PluginManager
import org.bukkit.plugin.java.JavaPlugin
import tr.s42.paper.PaperPlatform
import tr.s42.spigot.SpigotPlatform

class TradeCyclePlugin  : JavaPlugin() {

    override fun onEnable() {
        checkForUpdates()
        saveDefaultConfig()

        val platform: Platform = getPlatform()
        val service = TradeCycleService(platform.messageService, config)
        registerStrategies(platform, service, config)
    }

    private fun checkForUpdates() {
        VersionCheckTask { latestVersion ->
            val currentVersion = ComparableVersion(description.version)
            if (currentVersion < latestVersion) {
                logger.warning("A new version of TradeCycle is available: ${latestVersion.rawVersion}. You are using version: ${currentVersion.rawVersion}")
                logger.warning("Download the latest version at: ${VersionCheckTask.RESOURCE}")
                return@VersionCheckTask
            }
        }.runTaskAsynchronously(this)
    }

    private fun getPlatform(): Platform {
        try {
            Class.forName("com.destroystokyo.paper.PaperConfig")
            return PaperPlatform()
        } catch (e: ClassNotFoundException) {
            return SpigotPlatform()
        }
    }

    private fun registerStrategies(platform: Platform, service: TradeCycleService, config: FileConfiguration) {
        val strategies: List<Listener> = config.getStringList("strategy").map {
            strategies[it] ?: throw IllegalArgumentException("Invalid strategy: $it")
        }

        val pluginManager: PluginManager = server.pluginManager
        pluginManager.registerEvents(VillagerCycleListener(service, platform.openInventoryViewFunction, this), this)
        strategies.forEach { pluginManager.registerEvents(it, this) }
    }
}